### DISCUSSION TITLE
The differentiability of translation in affine transformation?

### DISCUSSION THREAD

**Discussion by JeffersonQin** [kornia/kornia#2771] [Questions]
Hi, I am trying to see if kornia's `affine()` is differentiable. The main code I wrote is as the following:

```python
class AffineTransform(nn.Module):
    def __init__(self):
        super(AffineTransform, self).__init__()
        self.theta = nn.Parameter(torch.tensor([[[1, 0, 0], [0, 1, 0]]], dtype=torch.float32))
        self.s = nn.Parameter(torch.tensor([[0.0, 0.0]], dtype=torch.float32)) # not necessary

    def forward(self, x):
        # grid = F.affine_grid(self.theta, x.size())
        # x = F.grid_sample(x, grid)
        x2 = kornia.geometry.transform.affine(x, self.theta)
        x3 = kornia.geometry.translate(x2, self.s) # not necessary
        return x3
```

This is invoked by a training loop, the code is used to do a registration,

The code in the comment is the affine function provided by torchvision.

The result I got is quite surprising, e.g.

![image](https://github.com/kornia/kornia/assets/30655701/d06bc61c-5072-424f-9043-11a5043172ca)

Also, it seems no matter how I try to change the optimization goal or even add a translation code, the left up corner always will not move. Thus I am wondering if translation is not differentiable in this case. Thank you so much for your guys help.

**Comment by edgarriba** [comment]
f you want learn the translation, be aware that's already part of the affine matrix transform ((tx, ty) == m[..., :, 2]). So i'm not sure why are you adding the extra translation step here. About the affine operator, the transform matrix is in pixel space (you can see that internally implements the [`warp_affine`](https://github.com/kornia/kornia/blob/main/kornia/geometry/transform/imgwarp.py#L131) ops, while when you use affine_grid, it's already normalized to work with ndc (normalized device coordinates == [-1, 1]).

In order to use properly kornia functions you might have to normalize/denormalize the transform matrix depending what you want to do (we have functionality for that too).  In any case, can you share a colab or something minimal to reproduce ? That could be also good material for tutorial. We can fix things in the api to make it more accessible for others. I saw also in the past similar things
